### ISSUE TITLE
Questions about batch_norm_on_an_extra_axis

### ISSUE THREAD

**Issue by BritaryZhou** [google-research/deeplab2#45]
Hi,

Thanks for your great work~
I just have some questions about the "batch_norm_on_an_extra_axis" function in the "max_deeplab.py" file. I am wondering about how much this operation affects the results.  And could you please have a check that whether the following Pytorch-based implementation is able to achieve the same effect as the current deeplab2 version? Thanks a lot~

```
self.bn = nn.BatchNorm2d(1)

batch_size, slot_num = pixel_space_mask_logits.size(0), pixel_space_mask_logits.size(1)
pixel_space_mask_logits = self.bn(rearrange(pixel_space_mask_logits, "b l h w -> (b l) a h w", a=1))
pixel_space_mask_logits = rearrange(pixel_space_mask_logits, "(b l) a h w -> b l h w", b=batch_size, l=slot_num)
```

Best Regards,

**Comment by csrhddlam** [comment]
Hi,

Thanks for your interest in our work.

1. The use of "batch_norm_on_an_extra_axis" can avoid careful scaling of the last layer which is the multiplication of an L2-normalized pixel space feature and a batch-normalized transformer kernel. This choice is simple and works well in practice.
2. Without the batch norm but with a careful choice of the scaling factor and a good learning rate: We didn't try this option, but I think it can probably work as well as the default batch norm choice.
3. Without the batch norm or a proper scaling factor, the output scale or the multiplication is not controlled and might not work as well as the previous two choices.

The PyTorch implementation looks correct to me. But note that we use synchronized batch norm for all segmentation cases.

Best,
Huiyu

**Comment by aquariusjay** [comment]
Closing the issue, and please feel free to reopen or create a new one if you still have any question.
